### ISSUE TITLE
Garbage collection

### ISSUE THREAD

**Issue by kendonB** [wlandau-lilly/drake#118]
Does drake have garbage collection; as in, can I run a command like:

```
drake_gc(my_plan)
```
And drake will delete all the files not associated with the targets in my_plan?

Same issue as https://github.com/richfitz/remake/issues/55.

**Comment by wlandau-lilly** [comment]
Just to be clear, do you and @richfitz mean just clearing out any dangling old targets in the cache? Because that's straightforward with `clean()`.

```r
load_basic_example()
make(my_plan)
cached()
```

```r
## [1] "coefficients"           "coef_regression1_large" "coef_regression1_small" "coef_regression2_large"
##  [5] "coef_regression2_small" "data.frame"             "knit"                   "large"                 
##  [9] "lm"                     "reg1"                   "reg2"                   "regression1_large"     
## [13] "regression1_small"      "regression2_large"      "regression2_small"      "'report.md'"           
## [17] "'report.Rmd'"           "rpois"                  "simulate"               "small"                 
## [21] "stats::rnorm"           "summary"                "summ_regression1_large" "summ_regression1_small"
## [25] "summ_regression2_large" "summ_regression2_small" "suppressWarnings"  
```

```r
new_plan <- workplan(a = 1, b = 2, result = c(a, b))
make(new_plan)
cached() # too many targets
```

```r
 ## [1] "a"                      "b"                      "c"                      "coefficients"          
##  [5] "coef_regression1_large" "coef_regression1_small" "coef_regression2_large" "coef_regression2_small"
##  [9] "data.frame"             "knit"                   "large"                  "lm"                    
## [13] "reg1"                   "reg2"                   "regression1_large"      "regression1_small"     
## [17] "regression2_large"      "regression2_small"      "'report.md'"            "'report.Rmd'"          
## [21] "result"                 "rpois"                  "simulate"               "small"                 
## [25] "stats::rnorm"           "summary"                "summ_regression1_large" "summ_regression1_small"
## [29] "summ_regression2_large" "summ_regression2_small" "suppressWarnings"    
```


```r
remove_these <- setdiff(cached(), new_plan$target)
file.exists("report.md") # TRUE
clean(list = remove_these)
file.exists("report.md") # FALSE (Custom file targets are removed when their hashes are uncached.)
cached()
```

```r
## [1] "a"      "b"      "result"
```

Do you think this is enough functionality? There is also a function called `prune()`, but it has been deprecated for months, and I haven't looked at it in about as long. And if you also wanted to remove files that were never part of any workflow plan, such as adjacent READMEs, things could get trickier.

**Comment by wlandau-lilly** [comment]
FYI: `imported()` lists all the cached imports (see also `built()`), so you could easily choose to not remove them. Imports are always re-processed with every `make()`, so removing them is usually not a big deal. But for large imported files, you might want to keep the hashes in the cache. That way, `drake` will use the file modification times to check for changes (see the first comment in #4).

**Comment by wlandau-lilly** [comment]
Closing because I think `clean()` does the job, as I explain above. @kendonB, feel free to comment and/or reopen if you disagree.

**Comment by richfitz** [comment]
FWIW, the "garbage collection" referred to in richfitz/remake#55 refers to the storr, not any files in the working copy or the cache of objects - every piece of data added to a storr stays there, even after deletion.  This is because multiple keys may point at the same bit of data.  There is no explicit reference counting so storr includes a `gc` method and finds all dangling bits of data and deletes them.  I think this is quite different to what @kendonB is concerned about here (but possibly a separate issue worth considering...)

**Comment by wlandau-lilly** [comment]
Sorry @richfitz, it took me a while to see this comment. It would be great if `storr` caches had garbage collection functionality themselves (see richfitz/storr#57).

**Comment by richfitz** [comment]
Yes, this is implemented in storr already - you just have to use it

**Comment by wlandau-lilly** [comment]
Great!

**Comment by wlandau-lilly** [comment]
Related to #154. Need to call `cache$gc()` in `clean()`, with the option to turn it off to go fast.

**Comment by wlandau-lilly** [comment]
Now available: `clean(garbage_collection = TRUE)`. Note: garbage collection could be mildly slow, so `garbage_collection` is `FALSE` by default.

**Comment by wlandau-lilly** [comment]
Also available: `drake_gc()` (effective from 07d871e501b0fb34e3cdf174747c654a103e5ce2). No need to `clean()` to have garbage collection.
